### PR TITLE
Adds NegativeInt scalar

### PR DESCRIPTION
**Pull request by mbertelsen**

### CODE REVIEW

File: src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace HotChocolate.Types
+{
+    /// <summary>
+    /// The NegativeIntType scalar type represents Integers that will have a value less than 0.
+    /// </summary>
+    public class NegativeIntType : IntType
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NegativeIntType "/> class.
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:13]:
```suggestion
        /// Initializes a new instance of the <see cref="NegativeIntType"/> class.
```

Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:13]:
```suggestion
        /// Initializes a new instance of the <see cref="NegativeIntType"/> class.
```

File: src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        /// </summary>
+        public NegativeIntType()
+            : this(
+                WellKnownScalarTypes.NegativeInt,
+                ScalarResources.NegativeIntType_Description)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NegativeIntType "/> class.
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:23]:
```suggestion
        /// Initializes a new instance of the <see cref="NegativeIntType"/> class.
```

File: src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            BindingBehavior bind = BindingBehavior.Explicit)
+            : base(name, description, int.MinValue, -1, bind)
+        {
+            Description = description;
+        }
+
+        /// <inheritdoc />
+        protected override bool IsInstanceOfType(int runtimeValue)
+        {
+            return runtimeValue < 0;
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:37]: You have access to the protected `MaxValue` :) 
```suggestion
            return runtimeValue <= MaxValue;
```

File: src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        /// <inheritdoc />
+        protected override bool IsInstanceOfType(IntValueNode valueSyntax)
+        {
+            return valueSyntax.ToInt32() < 0;
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:43]:
```suggestion
            return valueSyntax.ToInt32() <= MaxValue;
```

File: src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        /// <inheritdoc />
+        protected override int ParseLiteral(IntValueNode valueSyntax)
+        {
+            if (valueSyntax.ToInt32() >= 0)
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:49]:
```suggestion
            if (valueSyntax.ToInt32() > MaxValue)
```

File: src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                throw ThrowHelper.NegativeIntType_ParseLiteral_IsNotNegative(this);
+            }
+
+            return base.ParseLiteral(valueSyntax);
+        }
+
+        /// <inheritdoc />
+        protected override IntValueNode ParseValue(int runtimeValue)
+        {
+            if (runtimeValue >= 0)
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/NegativeIntType.cs:60]:
```suggestion
            if (runtimeValue > MaxValue)
```

File: src/HotChocolate/Core/src/Types.Scalars/ScalarResources.Designer.cs [hunk 7]
```diff
@@ -11,32 +12,46 @@ namespace HotChocolate.Types.Scalars {
...
-    [System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "4.0.0.0")]
-    [System.Diagnostics.DebuggerNonUserCodeAttribute()]
-    [System.Runtime.CompilerServices.CompilerGeneratedAttribute()]
+    /// <summary>
+    ///   A strongly-typed resource class, for looking up localized strings, etc.
+    /// </summary>
+    // This class was auto-generated by the StronglyTypedResourceBuilder
+    // class via a tool like ResGen or Visual Studio.
+    // To add or remove a member, edit your .ResX file then rerun ResGen
+    // with the /str option, or rebuild your VS project.
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/ScalarResources.Designer.cs:21]: Interesting. Are you using Visual Studio? 
@michaelstaib  how do we handle different formats of the resources depending on the IDE?

File: src/HotChocolate/Core/src/Types.Scalars/ScalarResources.resx [hunk 8]
```diff
@@ -1,23 +1,131 @@
...
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
+  <data name="NegativeIntType_Description" xml:space="preserve">
+    <value>The NegativeIntType scalar type represents Integers that will have a value less than 0.</value>
+  </data>
+  <data name="NegativeIntType_IsNotNegative_ParseLiteral" xml:space="preserve">
+    <value>NegativeIntType cannot parse the provided literal. The provided int was not negative.</value>
+  </data>
+  <data name="NegativeIntType_IsNotNegative_ParseValue" xml:space="preserve">
+    <value>NegativeIntType cannot parse the provided value. The provided int was not Negative.</value>
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Types.Scalars/ScalarResources.resx:127]:
```suggestion
    <value>NegativeIntType cannot parse the provided value. The provided int was not negative.</value>
```

File: src/HotChocolate/Core/test/Types.Scalars.Tests/NegativeIntTypeTests.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+            // act
+            // assert
+            schema.ToString().MatchSnapshot();
+        }
+
+        [Theory]
+        [InlineData(typeof(EnumValueNode), TestEnum.Foo, false)]
+        [InlineData(typeof(FloatValueNode), 1d, false)]
+        [InlineData(typeof(IntValueNode), -1, true)]
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Types.Scalars.Tests/NegativeIntTypeTests.cs:24]: It would be nice to see the edge cases also tested
```suggestion
        [InlineData(typeof(IntValueNode), -1, true)]
        [InlineData(typeof(IntValueNode), int.MinValue, true)]
        [InlineData(typeof(IntValueNode), 0, false)]
```

File: src/HotChocolate/Core/test/Types.Scalars.Tests/NegativeIntTypeTests.cs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+            // act
+            // assert
+            ExpectIsInstanceOfTypeToMatch<NegativeIntType>(valueNode, expected);
+        }
+
+        [Theory]
+        [InlineData(TestEnum.Foo, false)]
+        [InlineData(1d, false)]
+        [InlineData(-1, true)]
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Types.Scalars.Tests/NegativeIntTypeTests.cs:45]: Same as above. Edge cases

File: src/HotChocolate/Core/test/Types.Scalars.Tests/NegativeIntTypeTests.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        [InlineData(true)]
+        [InlineData("")]
+        public void ParseLiteral_GivenObject_ThrowSerializationException(object value)
+        {
+            // arrange
+            // act
+            // assert
+            ExpectParseValueToThrowSerializationException<NegativeIntType>(value);
+        }
+    }
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Types.Scalars.Tests/NegativeIntTypeTests.cs:116]: I forgot something in my pr :) 

We also need to test the serialisation and deserialisation

For the positive int I used these test cases:

```csharp

        [Theory]
        [InlineData(1, 1)]
        [InlineData(int.MaxValue, int.MaxValue)]
        [InlineData(null, null)]
        public void Deserialize_GivenValue_MatchExpected(
            object resultValue,
            object runtimeValue)
        {
            // arrange
            // act
            // assert
            ExpectDeserializeToMatch<PositiveIntType>(resultValue, runtimeValue);
        }

        [Theory]
        [InlineData(TestEnum.Foo)]
        [InlineData(1d)]
        [InlineData(true)]
        [InlineData("foo")]
        [InlineData(int.MinValue)]
        [InlineData(-1)]
        [InlineData(0)]
        public void Deserialize_GivenValue_ThrowSerializationException(object value)
        {
            // arrange
            // act
            // assert
            ExpectDeserializeToThrowSerializationException<PositiveIntType>(value);
        }

        [Theory]
        [InlineData(1, 1)]
        [InlineData(int.MaxValue, int.MaxValue)]
        [InlineData(null, null)]
        public void Serialize_GivenObject_MatchExpectedType(
            object runtimeValue,
            object resultValue)
        {
            // arrange
            // act
            // assert
            ExpectSerializeToMatch<PositiveIntType>(runtimeValue, resultValue);
        }

        [Theory]
        [InlineData(TestEnum.Foo)]
        [InlineData(1d)]
        [InlineData(true)]
        [InlineData("foo")]
        [InlineData(int.MinValue)]
        [InlineData(-1)]
        [InlineData(0)]
        public void Serialize_GivenObject_ThrowSerializationException(object value)
        {
            // arrange
            // act
            // assert
            ExpectSerializeToThrowSerializationException<PositiveIntType>(value);
        }

```

**Review by PascalSenn — CHANGES_REQUESTED** [review]
Hi @mbertelsen 

This looks already really good! well done :) 

I just have some minor things (some things I forgot in my pr too 👼 )

Also, one small thing missing, can you also add the docs?
website/src/docs/hotchocolate/defining-a-schema/scalars.md

Something like:  
 https://github.com/ChilliCream/hotchocolate/blob/78f760c8e629c9a3ed46a49e8459b5fcc1d574eb/website/src/docs/hotchocolate/defining-a-schema/scalars.md#positiveint

File: src/HotChocolate/Core/src/Types.Scalars/ScalarResources.Designer.cs (hunk 7 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types.Scalars/ScalarResources.Designer.cs:21]: I will have a look ....

Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types.Scalars/ScalarResources.Designer.cs:21]: This looks alright.

Its generated ... so as long as it works we ignore this code.

Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types.Scalars/ScalarResources.Designer.cs:21]: [global::System.Runtime.CompilerServices.CompilerGeneratedAttribute()]
 
 sonar will ship code that is generated

**Review by PascalSenn — APPROVED** [review]
LGTM

**Comment by PascalSenn** [comment]
@mbertelsen  Thanks for the contribution 🚀 

We will merge the feat/scalars branch once it is ready 👍
